### PR TITLE
Longer cluster name length + keep cluster name prefix

### PR DESCRIPTION
**Pull request by ijsnellf**
**What this PR does / why we need it**: The max cluster name length is now configurable in Envoy. This PR updates the max cluster name from 60 to 189 characters and only generates a SHA if necessary. When a cluster name is longer than 189 characters a SHA is generated. The first 149 characters of the cluster name is prefixed with the original cluster name and the remaining 40 characters are the SHA of the original.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Changed the max cluster name length to 189 characters from 60 in Envoy.
```

### CODE REVIEW

File: pilot/proxy/envoy/config.go [hunk 1]
```diff
@@ -792,6 +792,15 @@ func appendPortToDomains(domains []string, port int) []string {
 	return domainsWithPorts
 }
 
+func truncateClusterName(name string) string {
+	if len(name) > MaxClusterNameLength {
+		prefix := name[:MaxClusterNameLength - sha1.Size * 2]
```
Comment (kyessenov) [inline pilot/proxy/envoy/config.go]: why multiply by a factor of 2?

Comment (ijsnellf) [reply pilot/proxy/envoy/config.go]: sha1.Size is in bytes. `fmt.Printf("%x", ...)`  encodes data to base16 so there are two chars per byte.

**Comment by kyessenov** [comment]
I don't have permissions to approve, but LGTM.

File: .gitignore [hunk 2]
```diff
@@ -15,4 +15,4 @@ cluster.env
 kubedns
 *.pem
 # pilot
-pilot/platform/config
+pilot/platform/config
```
Comment (ldemailly) [inline .gitignore]: please add back the missing newline / don't change this file

File: WORKSPACE [hunk 3]
```diff
@@ -1048,7 +1048,7 @@ go_repository(
 
 # Change this and the docker/Dockerfile.proxy* files together
 # This SHA is obtained from proxy/postsubmit job
-ISTIO_PROXY_BUCKET = "1509b9361c3e727b40146dcddd44ed0012b7de78"
+ISTIO_PROXY_BUCKET = "a05c813aa01b8eed00d1698f5daba3f468b0a293"
```
Comment (ldemailly) [inline WORKSPACE]: those changes are supposed to happen automatically with the bot (but maybe the bot is down?)

File: pilot/proxy/envoy/config.go [hunk 4]
```diff
@@ -792,6 +792,15 @@ func appendPortToDomains(domains []string, port int) []string {
 	return domainsWithPorts
 }
 
+func truncateClusterName(name string) string {
```
Comment (ldemailly) [inline pilot/proxy/envoy/config.go]: can you add a unit test with boundary checks (ie size just under max, at max, at max + 1)

Comment (ijsnellf) [reply pilot/proxy/envoy/config.go]: yep!

File: WORKSPACE (hunk 3 above)
Comment (ijsnellf) [reply WORKSPACE]: yes. it is probably not operating yet after the mono-repo reshuffle.

File: pilot/proxy/envoy/testdata/cds-circuit-breaker.json.golden [hunk 5]
```diff
@@ -100,26 +99,13 @@
     }
    },
    {
-    "name": "out.12a577b4dd04c9a3fed313fc71fb850e7f9c345f",
-    "service_name": "world.default.svc.cluster.local|redis",
-    "connect_timeout_ms": 1000,
-    "type": "sds",
```
Comment (rshriram) [inline pilot/proxy/envoy/testdata/cds-circuit-breaker.json.golden]: ? why nuke?

Comment (ijsnellf) [reply pilot/proxy/envoy/testdata/cds-circuit-breaker.json.golden]: it isn't nuked. my change altered the ordering and it got moved down here: https://github.com/istio/istio/pull/1344/files#diff-99b71f016ec8a2c4cc3db2e721464b3dR265

**Comment by ijsnellf** [comment]
@rshriram: my change modified the cluster names which in turn changed the ordering. Unfortunately this means the diff is quite ugly but the output is unchanged aside from cluster names and cluster order.

**Comment by ijsnellf** [comment]
I think it is because we use the cluster name as a key here:
https://github.com/istio/istio/blob/master/pilot/proxy/envoy/resources.go#L673

**Pull request merged**